### PR TITLE
fix(schematics): ngrx schematics should generate enhanced ngrx files

### LINKED ISSUE #317: Auto create file for selectors. => `app.selectors.ts`
I would suggest creating a file called `app.selectors.ts` automatically with the rest of the files in `+state`.
Now we can group and reuse the selector logic and organize it together with actions, reducer, and effects.

Example:
```typescript
// app/+state/app.selectors.ts

import {createSelector} from '@ngrx/store';
import {App, AppState} from './app.interfaces';

export function getAppState(s: AppState): App {
  return s.app
}

export const getCount = createSelector(
  getAppState,
  (state: App) => state.count
)
```

```typescript
// app/path/to/my.component.ts
import {getCount} from './+state/app.selectors';
...

count$: Observable<number>;
...
constructor(private s: DataPersistence<AppState>) {
    ...
    this.count$ = this.s.store.select(getCount);
}

```

### LINKED ISSUE #472: feat(ngrx): properly scaffold  Load/Loaded/LoadError actions for each feature
Recent enhancements to the [ngrx/platform schematics](https://github.com/ngrx/platform/pull/1047 ) creates more specific *feature* actions and effects classes. 

As such, in the *Nx ngrx* schematics, enhancements should be insure that the following actions are scaffolded:

* For each ngrx `name` feature, the following three (3) actions should be prepared:
   ```ts
     export type <%= classify(name) %>Actions = 
            Load<%= classify(name) %>s     | 
            <%= classify(name) %>sLoaded | 
            <%= classify(name) %>LoadError;
   ```

### LINKED ISSUE #561: ng g ngrx seems broken in 6.0
After generating auth in libs:
```
ng g ngrx auth --module=libs/auth/src/lib/auth.module.ts
```

> Path "/libs/auth/src/lib/+state/auth.effects.spec.ts" does not exist.

### LINKED ISSUE #618: ng g ngrx doesn't create all files
Hi
There is something not clear regarding which files should be generated for ngrx.
Currently when I use that functionality, only actions, effects, reducer files are created but documentation says that also interfaces.ts and init.ts files should be. See: https://github.com/nrwl/nx/blob/68e34d494fdca1c57a757c889ea8105e8a0a7a6e/packages/schematics/src/collection/ngrx/ngrx.md

There is also merged PR: https://github.com/nrwl/nx/pull/386 which states:

> ...
> generate init and interface files
> ...
> 

but in the commit I can see that generation of these files was removed.

### PR DESCRIPTION
**Pull request by ThomasBurleson**
@nrwl/schematics no longer uses the @ngrx/schematics to generate NgRx feature files.
*  `ngrx/files/__directory__` templates are used
*  Templates replicate the simple outputs generated from @ngrx/schematics:feature
*  Templates add significant Nx enhancements.

The following standard files will be scaffolded:
* `<feature>.actions.ts`
* `<feature>.effects.ts` + `<feature>.effects.spec.ts`
* `<feature>.reducer.ts` + `<feature>.reducer.spec.ts`

The following new files will also be scaffolded:
* `<feature>.interfaces.ts`
* `<feature>.selectors.ts` +  `<feature>.selectors.spec.ts`
* `<feature>.facade.ts` + `<feature>.facade.spec.ts`

Changes include:

* Change the action/enums to generate a trio of enums for each *feature*: `Load<Feature>`, `<Feature>Loaded`, and `<Feature>LoadError`
* Add code generators for `<feature>.selectors.ts`, `<feature>.interfaces.ts`, `<feature>.facade.ts`
* Add code generators for unit and integration testing `*.spec.ts` files
* Update the public barrel [`index.ts`] when adding ngrx to a library
* Use `StoreModule.forFeature()` when adding ngrx feature (without using the `--root` option)
* Use the Effect to respond tp `load<Feature>$` and dispatch `<Feature>Loaded` or `<Feature>LoadError`
* Update the Action to export `<feature>Actions` map of all action classes

Fixes #472,  Fixes #618,  Fixes #317,  Fixes #561, Refs #380.

### CODE REVIEW

**Comment by vsavkin** [comment]
@ThomasBurleson build fails

**Comment by ThomasBurleson** [comment]
@vsavkin, @FrozenPandaz - This is ready for review. 

Note we no longer the @ngrx/schematics to generate the actions, reducer, and effects classes. 

Those @ngrx/schematic generated ones were so simple and need extensive enhancements to meet our own Nx ngrx goals. For that reason (and others), it made sense to use our own templates which extended ideas from the latest @ngrx templates (see below for sample outputs).

Our version builds upon those AND includes scaffolding for interfaces, selectors, and facade classes.

> As `@ngrx/schematic` improve, we can easily monitor their changes and adjust accordingly.

----

####  ` @ngrx/schematics` sample output

* `<feature>.actions.ts`
```ts
import { Action } from '@ngrx/store';

export enum UserActionTypes {
  LoadUsers = '[User] Load Users'
}

export class LoadUsers implements Action {
  readonly type = UserActionTypes.LoadUsers;
}

export type UserActions = LoadUsers;
```

* `<feature>.reducer.ts`
```ts
import { Action } from '@ngrx/store';
import { UserActions, UserActionTypes } from './user.actions';

export interface State {

}

export const initialState: State = {

};

export function reducer(state = initialState, action: UserActions): State {
  switch (action.type) {

    case UserActionTypes.LoadUsers:
      return state;


    default:
      return state;
  }
}
```

* `<feature>.effect.ts`
```ts
import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { UserActionTypes } from './user.actions';
import { Observable } from 'rxjs';

@Injectable()
export class UserEffects {

  @Effect()
  loadFoos$ : Observable<any> = this.actions$.pipe(ofType(UserActionTypes.LoadUsers));

  constructor(private actions$: Actions) {}
}
```

**Comment by ankemp** [comment]
This should also fix #561?

**Comment by ThomasBurleson** [comment]
Ready for Review.

**Comment by ThomasBurleson** [comment]
I added the last tests for Facades.

File: scripts/test_schematics.sh [hunk 1]
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env bash
 
+# npm run test:schematics schematics/src/collection/ngrx/ngrx
```
Comment (vsavkin) [inline scripts/test_schematics.sh]: should be removed

File: packages/schematics/src/utils/testing-utils.ts [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
+  barrel: string;
+}
 
 var appConfig: AppConfig; // configure built in createApp()
+var libConfig: LibConfig;
 
 export function getAppConfig(): AppConfig {
   return appConfig;
 }
+export function getLibConfig() {
```
Comment (vsavkin) [inline packages/schematics/src/utils/testing-utils.ts]: add the return type for consistency

File: packages/schematics/src/utils/name-utils.ts [hunk 3]
```diff
@@ -40,3 +43,16 @@ function toCapitalCase(s: string): string {
 export function findModuleParent(modulePath) {
   return path.dirname(modulePath);
 }
+
+/**
+ * Simple logic to ensure the name is a 'plural'.
+ * @see https://github.com/blakeembrey/pluralize/blob/master/pluralize.js for complex pluralizations
+ */
+export function pluralize(name: string): string {
```
Comment (vsavkin) [inline packages/schematics/src/utils/name-utils.ts]: isn't used. remove it.

in general, autopluralizers are problematic (as say Rails showed)

File: packages/schematics/src/utils/name-utils.ts [hunk 4]
```diff
@@ -22,6 +24,7 @@ export function toPropertyName(s: string): string {
     .replace(/^([A-Z])/, m => m.toLowerCase());
 }
 
+// Upper camelCase to lowercase, hypenated
```
Comment (vsavkin) [inline packages/schematics/src/utils/name-utils.ts]: use /** */ format

File: packages/schematics/src/utils/name-utils.ts [hunk 5]
```diff
@@ -9,10 +9,12 @@ export function names(name: string): any {
   };
 }
 
+// hypenated to UpperCamelCase
 export function toClassName(str: string): string {
   return toCapitalCase(toPropertyName(str));
 }
 
+// Hypenated to lowerCamelCase
```
Comment (vsavkin) [inline packages/schematics/src/utils/name-utils.ts]: same here

File: packages/schematics/src/collection/ngrx/schema.json [hunk 6]
```diff
@@ -6,34 +6,35 @@
...
+      "default": "+state",
+      "description":
+        "Override the name of the folder used to contain/group the NgRx files: contains actions, effects, reducers. selectors. (e.g., +state)"
+    },
     "module": {
       "type": "string",
+      "default": "",
       "description":
-        "Path to ngModule; host directory will contain the new '+state' directory (e.g., src/libs/mylib/mylib.module.ts)."
+        "Path to ngModule; host directory will contain the new '+state' directory (e.g., src/libs/comments/state/lib/comments-state.module.ts)."
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/schema.json]: this path won't exist an nx workspace. Use a real path (e.g., libs/comments/src/lib/state/comments-state.module.ts).

File: packages/schematics/src/collection/ngrx/rules/add-imports-to-module.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import { Rule, Tree } from '@angular-devkit/schematics';
-import { Change } from '@ngrx/schematics/src/utility/change';
+import { Change } from '@ngrx/schematics/schematics-core';
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/rules/add-imports-to-module.ts]: I'd not use this type.

It's used here.

```
  const addImport = (symbolName: string, fileName: string): Change => {
      return insertImport(source, modulePath, symbolName, fileName);
    };
```

insertImport is provided by Angular schematics, so it seems wrong to use a type provided by NgRx for a function call provided by Angular

File: packages/schematics/src/collection/ngrx/ngrx.md [hunk 8]
```diff
@@ -1,61 +1,80 @@
-# ngrx
---------
+# Angular CLI `@nrwl/schematics:NgRx`
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/ngrx.md]: I don't think we should have this doc here.

We have API docs deployed to `nrwl.io/nx`. This doc appears to server a similar purpose, but I don't think a lot of folks will be able to find it.

I'd remove this doc and update `nrwl.io/nx` instead.

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.selectors.ts__tmpl__ [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import { createFeatureSelector, createSelector } from '@ngrx/store';
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.selectors.ts__tmpl__]: What does the ngrx/schematics package generates for selectors? Something similar?

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.reducer.ts__tmpl__ [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import { <%= className %>, <%= className %>State } from './<%= fileName %>.interfaces';
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.reducer.ts__tmpl__]: I thought we decided to remove interfaces to be closer to the default ngrx schematics.

Did they add the .interfaces file?

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.facade.ts__tmpl__ [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+import { Actions } from '@ngrx/effects';
+
+import { <%= className %>State } from './<%= fileName %>.interfaces';
+import { <%= propertyName %>Query } from './<%= fileName %>.selectors';
+import { Load<%= className %> } from './<%= fileName %>.actions';
+
+@Injectable({
+  providedIn : 'root'
+})
+export class <%= className %>Facade {
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.facade.ts__tmpl__]: NgRx doesn't have a notion of the facade, so I don't think we should generate it by default. What we generate by default should be as closer to the default ngrx stuff as possible.

Could you remove everything facade related from this PR? Let's talk about it more and maybe add it separately as an opt-in thing.

File: e2e/utils.ts [hunk 12]
```diff
@@ -65,6 +65,11 @@ export function copyMissingPackages(): void {
     `rm -rf tmp/${projectName}/node_modules/@angular-devkit/core/node_modules`
   );
 
+  execSync(`rm tmp/${projectName}/node_modules/.bin/semver`);
```
Comment (vsavkin) [inline e2e/utils.ts]: Why is this workaround required?

**Review by vsavkin — CHANGES_REQUESTED** [review]

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.selectors.ts__tmpl__ (hunk 9 above)
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.selectors.ts__tmpl__]: No selectors are generated in the current version. I believe Brandon has been playing with this idea.

File: e2e/utils.ts (hunk 12 above)
Comment (ThomasBurleson) [reply e2e/utils.ts]: `npm run e2e ngrx` was failing with *semver* errors. @FrozenPandaz  recommended this as the resolution.

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.effects.spec.ts__tmpl__ [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        <%= className %>Effects,
+        DataPersistence,
+        provideMockActions(() => actions)
+      ],
+    });
+
+    effects = TestBed.get(<%= className %>Effects);
+  });
+
+  describe('someEffect', () => {
```
[the remark this answers is not in the document]
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.effects.spec.ts__tmpl__]: Fixed.

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.effects.spec.ts__tmpl__ [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+import { EffectsModule } from '@ngrx/effects';
+import {StoreModule} from '@ngrx/store';
+import {provideMockActions} from '@ngrx/effects/testing';
+
+import { NxModule } from '@nrwl/nx';
+import {DataPersistence} from '@nrwl/nx';
+import {hot} from '@nrwl/nx/testing';
+
+
```
[the remark this answers is not in the document]
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.effects.spec.ts__tmpl__]: Fixed.

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.actions.ts__tmpl__ [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+ readonly type = <%= className %>ActionTypes.Load<%= className %>;
+ constructor(public payload: any) { }
+}
+
+export class <%= className %>Loaded implements Action {
+ readonly type = <%= className %>ActionTypes.<%= className %>Loaded;
+ constructor(public payload: any[]) { }
+}
+
+export type <%= className %>ActionsUnion = Load<%= className %> | <%= className %>Loaded | <%= className %>LoadError;
```
[the remark this answers is not in the document]
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.actions.ts__tmpl__]: Agreed. Fixed.

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.reducer.spec.ts__tmpl__ [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+       name: name || `name-${id}`
+     });
+  });
+
+  describe('valid <%= className %> actions ', () => {
+    it('should return set the list of known <%= className %>', () => {
+        const <%= propertyName %>s = [create<%= className %>( 'PRODUCT-AAA' ),create<%= className %>( 'PRODUCT-zzz' )];
+        const action = new <%= className %>Loaded(<%= propertyName %>s);
+        const result : <%= className %>State = <%= propertyName %>Reducer(initialState, action);
+        const selId : string = get<%= className %>Id(result.list[1]);
```
[the remark this answers is not in the document]
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.reducer.spec.ts__tmpl__]: When the spec is executed without this, the console with show TS2339 error. 
For example, within the generated spec file for `user` feature:

```console
ERROR in apps/myapp/src/app/+state/users.reducer.spec.ts(20,39): error TS2339: Property 'id' does not exist on type 'UsersState'.
```

File: packages/schematics/src/collection/ngrx/schema.json (hunk 6 above)
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/schema.json]: Fixed.

File: packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.selectors.spec.ts__tmpl__ [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+      const results = <%= propertyName %>Query.getAll<%= className %>(storeState);
+      const selId = get<%= className %>Id(results[1]);
+
+      expect(results.length).toBe(3);
+      expect(selId).toBe('PRODUCT-BBB');
+    });
+
+    it('getSelected<%= className %>() should return the selected <%= className %>', () => {
+      const result = <%= propertyName %>Query.getSelected<%= className %>(storeState);
+      const selId = get<%= className %>Id(result);
```
[the remark this answers is not in the document]
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/files/__directory__/__fileName__.selectors.spec.ts__tmpl__]: See [comment below](https://github.com/nrwl/nx/pull/474#discussion_r202956687)

File: packages/schematics/src/collection/ngrx/rules/add-imports-to-module.ts (hunk 7 above)
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/rules/add-imports-to-module.ts]: Fixed with `import { Change } from '@schematics/angular/utility/change';`

**Comment by ThomasBurleson** [comment]
@vsavkin, @FrozenPandaz - PR updated. Thx.

**Pull request merged**